File: src/edu/uf/interactable/MIP1B.java
package edu.uf.interactable;

import edu.uf.Diffusion.Diffuse;
import edu.uf.compartments.GridFactory;
import edu.uf.intracellularState.Phenotype;
import edu.uf.primitives.Interactions;
import edu.uf.utils.Constants;
import edu.uf.utils.Util;

public class MIP1B extends Chemokine{
    public static final String NAME = "MIP1B";
    public static final int NUM_STATES = 1;
    
    private static MIP1B molecule = null; 

    private MIP1B(double[][][][] qttys, Diffuse diffuse) {
        super(qttys, diffuse);
        Macrophage.setChemokine(MIP1B.NAME);
        this.setPhenotye(Phenotype.createPhenotype());
    }
    
    public static MIP1B getMolecule(Diffuse diffuse) {
    	if(molecule == null) {
    		double[][][][] values = new double[NUM_STATES][GridFactory.getXbin()][GridFactory.getYbin()][GridFactory.getZbin()];
    		molecule = new MIP1B(values, diffuse); 
    	}
    	return molecule;
    }
    
    public static MIP1B getMolecule() {
    	return getMolecule(null);
    }
    
    @Override
    public double getKd() {
    	return Constants.Kd_MIP1B;
    }
    
    public void degrade() {
    	degrade(Constants.MIP1B_HALF_LIFE, 0);
    }

    public int getIndex(String str) {
        return 0;
    }

    public void computeTotalMolecule(int x, int y, int z) {
    	this.totalMoleculesAux[0] = this.totalMoleculesAux[0] + this.get(0, x, y, z);
    }

    protected boolean templateInteract(Interactable interactable, int x, int y, int z) {
        if (interactable instanceof PneumocyteII) 
        	return Interactions.secrete((Cell) interactable, this, Constants.P_MIP1B_QTTY, x, y, z, 0);
        
        if (interactable instanceof Macrophage) 
        	return Interactions.secrete((Leukocyte) interactable, this, Constants.MA_MIP1B_QTTY, x, y, z, 0);
        
        return interactable.interact(this, x, y, z); 
    }

    public double chemoatract(int x, int y, int z) {
    	//System.out.println("MIP: " + this.values[0][x][y][z]);
        return Util.activationFunction(this.get(0, x, y, z), Constants.Kd_MIP1B, Constants.VOXEL_VOL, 1, Constants.STD_UNIT_T) + Constants.DRIFT_BIAS;
    }

	@Override
	public String getName() {
		return NAME;
	} 

	@Override
	public double getThreshold() {
		return -1;
	}

	@Override
	public int getNumState() {
		return NUM_STATES;
	}
	
	@Override
	public boolean isTime() {
		return true;
	}
    
}

File: src/edu/uf/interactable/MIP1B.cpp
/*
 * MIP1B.cpp
 *
 *  Created on: Jul 19, 2024
 *      Author: henriquedeassis
 */

#include "MIP1B.h"
#include "../utils/Id.h"
#include "../utils/Constants.h"
#include "../utils/Util.h"
#include "../primitives/Interactions.h"
#include "Macrophage.h"
#include "PneumocyteII.h"
#include <memory>

namespace edu {
namespace uf {
namespace interactable {

using namespace primitives;

const int MIP1B::classId = Id::getId();
const string MIP1B::NAME = "MIP1B";
const int MIP1B::NUM_STATES = 1;
MIP1B* MIP1B::molecule = nullptr;

MIP1B::MIP1B(double**** qttys, int xbin, int ybin, int zbin, char wbin, Diffuse* diffuse)
    : Chemokine(qttys, xbin, ybin, zbin, wbin, diffuse) {
    setPhenotype(Id::getId());
}

MIP1B* MIP1B::getMolecule(Diffuse* diffuse, int xbin, int ybin, int zbin, char wbin) {
    if (molecule == nullptr) {
        double**** values = new double***[wbin];
        for (int i = 0; i < wbin; ++i) {
            values[i] = new double**[xbin];
            for (int j = 0; j < xbin; ++j) {
                values[i][j] = new double*[ybin];
                for (int k = 0; k < ybin; ++k) {
                    values[i][j][k] = new double[zbin]();
                }
            }
        }
        molecule = new MIP1B(values, xbin, ybin, zbin, wbin, diffuse);
    }
    return molecule;
}

int MIP1B::getClassId() const {
	return classId;
}

double MIP1B::getKd() {
    return utils::constexprants::Kd_IL10;
}

void MIP1B::degrade() {
    //degrade(utils::constexprants::TNF_HALF_LIFE, 0);
}

int MIP1B::getIndex(const char* str) const {
    return 0;
}

void MIP1B::computeTotalMolecule(int x, int y, int z) {
    totalMoleculesAux[0] += get(0, x, y, z);
}

bool MIP1B::templateInteract(Interactable* interactable, int x, int y, int z) {
	if (dynamic_cast<PneumocyteII*>(interactable))
		return Interactions::secrete(static_cast<Cell*>(interactable), this, utils::constexprants::P_MIP2_QTTY, x, y, z, 0);
	if (dynamic_cast<Macrophage*>(interactable))
		return Interactions::secrete(static_cast<Cell*>(interactable), this, utils::constexprants::MA_MIP2_QTTY, x, y, z, 0);

    return interactable->interact(this, x, y, z);
}

string MIP1B::getName() const {
    return NAME;
}

double MIP1B::getThreshold() const {
    return 0;
}

int MIP1B::getNumState() const {
    return NUM_STATES;
}

bool MIP1B::isTime() const {
    return true;
}

double MIP1B::chemoatract(int x, int y, int z){
	return utils::Util::activationFunction(this->get(0, x, y, z), utils::constexprants::Kd_MIP1B, utils::constexprants::VOXEL_VOL, 1, utils::constexprants::STD_UNIT_T) + utils::constexprants::DRIFT_BIAS;
}

} /* namespace interactable */
} /* namespace uf */
} /* namespace edu */
